infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,28 +73,40 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheHash.get(conflictKey);
+    }
+
+    private long firstAddress() {
+        if (cacheEntries.peek() == null)
+            return Address.NOT_FOUND;
+        return cacheEntries.peek().txVersion;
     }
 
     /**
-     * Invalidate all records up to a trim mark.
+     * Invalidate the first record with the minAddress.
+     */
+    private void invalidateFirst() {
+        ConflictTxStreamEntry entry = cacheEntries.poll();
+        cacheHash.remove(entry.conflictTxStream);
```

Comment:
Looks like you can use `TreeMap` instead of `PriorityQueue + HashMap`
Please consider can we replace it and if we can't then why?

Reply:
@xnull A `TreeMap` cannot have duplicates, where as a `PriorityQueue` can. Although, its possible to use a tree map, but the modelling has to be changed. Yes, you can can a "free" range remove method, but will it be simpled? I don't know, maybe you should propose something.

Reply:
Good point, we can replace priorityque with treemap that will be much smaller size if the number of entries are big, it can save some place. Will change the logic accordingly.